Audit Logging Support on Sync Gateway

ElliotFrancisHunter opened the pull request:
This is a PR for Audit Logging Support in Sync Gateway.

Preview link: https://preview.docs-test.couchbase.com/sgwaudlog/home/index.html

Credentials: https://couchbasecloud.atlassian.net/wiki/spaces/DOCS/pages/1971224949/Docs+Team+demo+site+passwords

Changed Pages
-------------------

- Audit Events reference: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/audit-log-events.html
- Audit Logs: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/audit-logging.html
- Manage Audit Logs: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/manage-audit-logs.html
- Supported Environments: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/supported-environments.html
- Platform Compatibility: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/compatibility.html
- What's new: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/whatsnew.html
- Release notes: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/release-notes.html
- Admin REST API: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/rest-api-admin.html
- Eventing Compatibility: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/server-compatibility-eventing.html
- Command Line Options: https://preview.docs-test.couchbase.com/sgwaudlog/sync-gateway/current/command-line-options.html

----------------------

Pages still needing changes/updates
-----------------------

-----------------------

ElliotFrancisHunter committed:
DOC-11833: Update message to legacy. (#798)

ElliotFrancisHunter committed:
[BLOCKER FOR APP SERVICES SCOPES AND COLLECTIONS] Sync Gateway 3.1.8 Maintenance Release Notes (#796)

* DOC-12233: Out of date file removal

* DOC-12233: SGW 3.1.7 relnotes updates

* DOC-12233: Dead file removal

* DOC-12233: Update relnote layout to match CBL relnotes

* Update sync-gateway-3-1-7-release-note.adoc

Adding extra bug fix

* DOC-12233: Release notes update

Update version, add some new tickets and update the xref

ElliotFrancisHunter committed:
Update antora.yml

Correct maintenance version missed from release

ElliotFrancisHunter committed:
DOC-12379: 3.1.9 relnotes (#802)

* DOC-12379: 3.1.9 relnotes

* Update sync-gateway-3-1-9-release-note.adoc

Removed ticket that's been shifted to a 3.1.10 release

ElliotFrancisHunter committed:
DOC-12304: Prep for 3.2.0 version

Change list:
-- Removing 3.1.x release notes from relevant pages
-- Removing 3.1.x content from What's New page and added dummy Audit Logging content
-- Added placeholder 3.2.0 release note
-- Update antora.yml to reflect version 3.2.0

ElliotFrancisHunter committed:
DOC-12304: Initial audit logging page implementation

Dummy content noting key points plus 'See Also' links

ElliotFrancisHunter committed:
DOC-12304: Audit log events ref page

Add audit events reference page.

ElliotFrancisHunter committed:
DOC-12304: Platform and compatibility

ElliotFrancisHunter committed:
DOC-12304: Add sgw beryllium

ElliotFrancisHunter committed:
Doc 12438 customer fixes (#804)

* DOC-12438: Customer fixes

Specify JS standard
Document CORS feedback.

* DOC-12438 update REST API specs

* add readOnly markup

* DOC-12438: Update legacy doc

* DOC-12438: Fix link formatting

* DOC-12348: Add link to Admin REST API

---------

Co-authored-by: Hakim Cassimally <hakim.cassimally@couchbase.com>

ElliotFrancisHunter committed:
Merge branch 'release/3.1' into DOC-12304-SyncGW-audit-logging

ElliotFrancisHunter committed:
DOC-12304: Release notes and small page tweaks

ElliotFrancisHunter committed:
DOC-12304: Audit logging concept doc v1

Also reorganised topics that fit better in the manage doc.
Fixed the Enterprise label so it no longer displays a URL.
Made the table formatting clearer.

torcolvin commented on modules/ROOT/pages/audit-logging.adoc:
```diff
@@ -0,0 +1,63 @@
+Viewing and managing audit logs permissions are enforced through Sync Gateway RBAC controls. 
+Logs are encrypted both in transit to prevent unauthorized access or tampering, all communication supports TLS.
+
+IMPORTANT: Rotated audit logs are not automatically pruned, administrators must manually configure audit log pruning.
```
These are pruned every 6 days, so actually what needs to happen is that the admin needs to make sure the rotated logs are picked up before they are deleted.

This value of 6 days might be wrong from a product perspective, I will need to followup with @idulo. That said, this is the value that is going out as default in 3.2.0 and should be documented as such.

ElliotFrancisHunter replied:
Changed

torcolvin commented on modules/ROOT/pages/audit-logging.adoc:
```diff
@@ -0,0 +1,63 @@
+
+Audit Logging provides tools for administrators to track operational irregularities and to support regulatory and security compliance standards, such as link:https://www.hhs.gov/hipaa/index.html[HIPAA] and link:https://soc2.co.uk/soc2[SOC-2].
+
+There is only one audit log file active at a time, which is periodically rotated based on the user-configured `rotateInterval` or `rotateSize` properties.
```
This will be:

- `logging.audit.rotation.max_age`
- `logging.audit.rotation.max_size`
- `logging.audit.rotation.rotated_logs_size_limit`
- `logging.audit.rotation.rotation_interval`

These match the rotation settings for any other logger e.g. https://docs.couchbase.com/sync-gateway/current/configuration-schema-bootstrap.html#logging-debug-rotation-max_age

torcolvin commented on modules/ROOT/pages/audit-logging.adoc:
```diff
@@ -0,0 +1,63 @@
+Audit Logging provides tools for administrators to track operational irregularities and to support regulatory and security compliance standards, such as link:https://www.hhs.gov/hipaa/index.html[HIPAA] and link:https://soc2.co.uk/soc2[SOC-2].
+
+There is only one audit log file active at a time, which is periodically rotated based on the user-configured `rotateInterval` or `rotateSize` properties.
+Logs are immutable and only accessible via read-only APIs, thus ensuring logs cannot be altered once recorded.
```
There's no API to access the logs for sync gateway. There might be for app services. The logs are immutable, however.

ElliotFrancisHunter replied:
Changed, thanks

torcolvin commented on modules/ROOT/pages/audit-logging.adoc:
```diff
@@ -0,0 +1,63 @@
+There is only one audit log file active at a time, which is periodically rotated based on the user-configured `rotateInterval` or `rotateSize` properties.
+Logs are immutable and only accessible via read-only APIs, thus ensuring logs cannot be altered once recorded.
+Viewing and managing audit logs permissions are enforced through Sync Gateway RBAC controls. 
+Logs are encrypted both in transit to prevent unauthorized access or tampering, all communication supports TLS.
```
There's no transport of sync gateway logs, they are written locally on each Sync Gateway node. *There is transport in app services.

ElliotFrancisHunter replied:
Changed, thanks!

osfameron committed:
DOC-12304 update REST API for 3.2 GA

From `3.2` tag in sync_gateway
8a3c0b46f88c4e20cef59c01ffe9ee015977b1d4 (HEAD, tag: 3.2, origin/release/3.2.1)

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
Not something that can be fixed within the timescale for this PR, but what is the difference between this file and `rest-api-admin.yaml` in the same directory? And why are these being maintained here, when there is also [admin-capella.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/admin-capella.yaml), [admin.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/admin.yaml), and [openapi.external.admin.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/openapi.external.admin.yaml) over in the `docs-capella` repo? Where is the single source of truth?

cc @osfameron , @torcolvin

torcolvin replied:
The canonical source of truth is in the Sync Gateway repo - tagged `3.2` (or major.minor for future releases) - there are different targets that munge the results for https://github.com/couchbase/sync_gateway/blob/main/.redocly.yaml

All of the changes that are in the the openapi specs can be made in the Sync Gateway repo, I can pick them up in a PR now though in future docs team are also welcome to create a ticket/PR for them.

There's a question for @osfameron whether docs should copy in the generated files to this repo. The files are unbundled in sync gateway repo.

ElliotFrancisHunter replied:
Just putting it here for posterity. I won't be addressing changes required in the bundled files. These are issues dragged from a generated spec so I imagine they'll be overwritten next time we pull/build the spec.

torcolvin replied:
https://issues.couchbase.com/browse/CBG-4204 for addressing these. I think in a time of less stress, that @ElliotFrancisHunter could (should?) take on doing this in the sync_gateway repo.

simon-dew commented on modules/ROOT/assets/attachments/bundled-metric.yaml:
As above, what is the difference between this file and `rest-api-metric.yaml` in the same directory? And why are these being maintained here, when there is also [metric-capella.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/metric-capella.yaml), [metric.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/metric.yaml), and [openapi.external.metric.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/openapi.external.metric.yaml) over in the `docs-capella` repo? Where is the single source of truth?

cc @osfameron, @torcolvin

simon-dew commented on modules/ROOT/assets/attachments/bundled-public.yaml:
As above, what is the difference between this file and `rest-api-public.yaml` in the same directory? And why are these being maintained here, when there is also [public-capella.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/public-capella.yaml), [public.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/public.yaml), and [openapi.external.public.yaml](https://github.com/couchbase/docs-capella/blob/main/modules/app-services/attachments/openapi.external.public.yaml) over in the `docs-capella` repo? Where is the single source of truth?

cc @osfameron, @torcolvin

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -67,7 +67,7 @@ paths:
             application/json:
               schema:
                 type: object
-                properties: &ref_108
```
Again, not something that can be fixed within this PR, but I think it is an anti-pattern to use YAML-specific anchors and aliases within an OpenAPI spec. (Although hand on heart, we've all done this.) YAML-specific anchors and aliases are not part of the OpenAPI spec and they have no equivalent in JSON. If you want to reuse content within an OpenAPI spec, the spec should be defined in chunks and assembled (e.g. using Redocly) via `$ref` fields. Something to look at for the future.

torcolvin replied:
https://github.com/couchbase/sync_gateway/blob/9a3f1b827990cd4d6bb1c2b7698ca87229f426e7/docs/api/paths/admin/db-_session.yaml#L15 is the canonical source of truth for this so this ref is generated by redocly bundle

simon-dew commented on modules/ROOT/assets/attachments/bundled-metric.yaml:
```diff
@@ -183,42 +183,72 @@ paths:
+                                description: HeapIdle is bytes in idle (unused) spans. Idle spans have no objects in them. These spans could be (and may already have been) returned to the OS, or they can be reused for heap allocations, or they can be reused as stack memory. HeapIdle minus HeapReleased estimates the amount of memory that could be returned to the OS, but is being retained by the ru ...
                               go_memstats_heapinuse:
                                 type: integer
+                                description: HeapInuse is bytes in in-use spans. In-use spans have at least one object in them. These spans an only be used for other objects of roughly the same size. HeapInuse minus HeapAlloc estimates the amount of memory that has been dedicated to particular size classes, but is not currently being used. This is an upper bound on fragmentation, but in general t ...
```
```suggestion
                                description: HeapInuse is bytes in in-use spans. In-use spans have at least one object in them. These spans can only be used for other objects of roughly the same size. HeapInuse minus HeapAlloc estimates the amount of memory that has been dedicated to particular size classes, but is not currently being used. This is an upper bound on fragmentation, but in general this memory can be reused efficiently.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-metric.yaml:
```diff
@@ -183,42 +183,72 @@ paths:
+                                description: The memory utilization (Resident Set Size) for the process, in bytes.
                               pub_net_bytes_recv:
                                 type: integer
+                                description: The total number of bytes received (since node start-up) on the network interface to which the Sync Gateway api.public_interface is bound. By default, that is the number of bytes received on 127.0.0.1:4984 since node start-up
```
```suggestion
                                description: The total number of bytes received (since node start-up) on the network interface to which the Sync Gateway api.public_interface is bound. By default, that is the number of bytes received on 127.0.0.1:4984 since node start-up.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                        default: 180
+                                        type: integer
+                              collation_buffer_size:
+                                description: The size of the log collation buffer
```
```suggestion
                                description: The size of the log collation buffer.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -1413,7 +1416,7 @@ paths:
                     **Behaviour**
                     * *default* - In priority order, this will cause
-                      - Deletes to always win (the delete with the logest revision history wins if both revisions are deletes)
+                      - Deletes to always win (the delete with the longest revision history wins if both revisions are deletes)
```
```suggestion
                      - Deletes to win always. (If both revisions are deletes, the delete with the longest revision history wins.)
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                    type: boolean
+                                    default: false
+                                  rotated_logs_size_limit:
+                                    description: Max Size (in mb) of log files before deletion
```
```suggestion
                                    description: Max Size (in mb) of log files before deletion.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                          - type: object
+                            properties: &ref_34
+                              enabled:
+                                description: Toggle for this log output
```
```suggestion
                                description: Toggle for this log output.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                        default: 6
+                                        type: integer
+                              collation_buffer_size:
+                                description: The size of the log collation buffer
```
```suggestion
                                description: The size of the log collation buffer.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                        default: 2
+                                        type: integer
+                              collation_buffer_size:
+                                description: The size of the log collation buffer
```
```suggestion
                                description: The size of the log collation buffer.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                        default: 2
+                                        type: integer
+                              collation_buffer_size:
+                                description: The size of the log collation buffer
```
```suggestion
                                description: The size of the log collation buffer.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2456,65 +2461,228 @@ paths:
+                                        default: 6
+                                        type: integer
+                              collation_buffer_size:
+                                description: The size of the log collation buffer
```
```suggestion
                                description: The size of the log collation buffer.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2539,6 +2707,21 @@ paths:
+                        description: Maximum number of concurrent replication connections allowed. If set to 0 this limit will be ignored.
+                        type: integer
+                      max_concurrent_changes_batches:
+                        description: Maximum number of changes batches to process concurrently per replication (1-5)"
```
```suggestion
                        description: Maximum number of changes batches to process concurrently per replication (1-5).
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2539,6 +2707,21 @@ paths:
+                        minimum: 1
+                        maximum: 5
+                      max_concurrent_revs:
+                        description: Maximum number of revs to process concurrently per replication (5-200)
```
```suggestion
                        description: Maximum number of revs to process concurrently per replication (5-200).
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2572,12 +2756,25 @@ paths:
                             type: boolean
+                            default: false
+                      allow_dbconfig_env_vars:
+                        description: Can be set to false to skip environment variable expansion in database configs
```
```suggestion
                        description: Can be set to false to skip environment variable expansion in database configs.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -2572,12 +2756,25 @@ paths:
+                        default: true
+                      diagnostic_interface:
+                        description: |-
+                          Network interface to bind diagnotic API to.
```
Not keen on the dangling preposition, but fix the typo anyway.

```suggestion
                          Network interface to bind diagnostic API to.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4173,6 +4387,46 @@ paths:
+                              - 1234
+                              - 5678
+                          disabled_users:
+                            description: List of users for which audit logging is disabled
```
```suggestion
                            description: List of users for which audit logging is disabled.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4173,6 +4387,46 @@ paths:
+                                  description: The name of the user for which audit logging is disabled.
+                                  type: string
+                          disabled_roles:
+                            description: List of roles for which audit logging is disabled
```
```suggestion
                            description: List of roles for which audit logging is disabled.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+
+        Required Sync Gateway RBAC roles:
+
+        * TODO
```
It's fair enough having a TODO in the file but comment it out so no risk of it being exposed to users.

```suggestion
      # TODO: Required Sync Gateway RBAC roles
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+            default: false
+      responses:
+        '200':
+          description: Successfully retrieved database configuration
```
```suggestion
          description: Successfully retrieved database configuration.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+
+        Required Sync Gateway RBAC roles:
+
+        * TODO
```
```suggestion
      # TODO: Required Sync Gateway RBAC roles
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+
+        * TODO
+      requestBody:
+        description: The new database audit configuration to use
```
```suggestion
        description: The new database audit configuration to use.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+                  properties: *ref_47
+      responses:
+        '200':
+          description: Database audit configuration successfully updated
```
```suggestion
          description: Database audit configuration successfully updated.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+        '200':
+          description: Database audit configuration successfully updated
+        '400':
+          description: There was a problem with your request
```
```suggestion
          description: There was a problem with your request.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+          description: There was a problem with your request
+          content: *ref_5
+        '404':
+          description: Resource could not be found
```
```suggestion
          description: Resource could not be found.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+
+        Required Sync Gateway RBAC roles:
+
+        * TODO
```
```suggestion
      # TODO: Required Sync Gateway RBAC roles
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+            type: boolean
+            default: false
+      requestBody:
+        description: The database configuration fields to update
```
```suggestion
        description: The database configuration fields to update.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+                  properties: *ref_47
+      responses:
+        '200':
+          description: Database audit configuration successfully updated
```
```suggestion
          description: Database audit configuration successfully updated.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -4301,6 +4555,167 @@ paths:
+        '200':
+          description: Database audit configuration successfully updated
+        '400':
+          description: There was a problem with your request
```
```suggestion
          description: There was a problem with your request.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -5879,42 +6294,72 @@ paths:
+                                description: The memory utilization (Resident Set Size) for the process, in bytes.
                               pub_net_bytes_recv:
                                 type: integer
+                                description: The total number of bytes received (since node start-up) on the network interface to which the Sync Gateway api.public_interface is bound. By default, that is the number of bytes received on 127.0.0.1:4984 since node start-up
```
```suggestion
                                description: The total number of bytes received (since node start-up) on the network interface to which the Sync Gateway api.public_interface is bound. By default, that is the number of bytes received on 127.0.0.1:4984 since node start-up.
```

simon-dew commented on modules/ROOT/assets/attachments/bundled-admin.yaml:
```diff
@@ -9061,74 +9519,90 @@ components:
+      additionalProperties: *ref_49
+      example: *ref_50
+    Resync-status:
+      description: The status of a resync operation
```
```suggestion
      description: The status of a resync operation.
```

simon-dew commented on modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-0-release-note.adoc:
```diff
@@ -0,0 +1,202 @@
+
+* https://issues.couchbase.com/browse/CBG-4003[CBG-4003 -  Fixed DeleteRole doesnt trigger history calculation for named collections]
+
+* https://issues.couchbase.com/browse/CBG-4015[CBG-4015 -  Fixed nextSequenceGreaterThan should update to current _sync:seq before releasing sequences]
```
Just be careful of underscores being interpreted as start of italics.

```suggestion
* https://issues.couchbase.com/browse/CBG-4015[CBG-4015 -  Fixed nextSequenceGreaterThan should update to current &lowbar;sync:seq before releasing sequences]
```

simon-dew commented on modules/ROOT/pages/_partials/sgw-svr-compatibility.adoc:
No indication of which versions of server are unsupported? Not a major issue

simon-dew commented on modules/ROOT/pages/_partials/static_restapi/admin/index.adoc:
Presumably generated from `bundled-admin.yaml`

simon-dew commented on modules/ROOT/pages/_partials/static_restapi/metric/index.adoc:
Presumably generated from `bundled-metric.yaml`

simon-dew commented on modules/ROOT/pages/_partials/static_restapi/public/index.adoc:
Presumably generated from `bundled-public.yaml`

simon-dew commented on modules/ROOT/pages/audit-logging.adoc:
```diff
@@ -0,0 +1,63 @@
+
+== About Audit Logging
+
+Key concepts, use cases and sign posting
+Explain audit log files, only one active at a time that they’re written etc
```
These look like notes to self. Comment out so there's no risk of them being shown to users

```suggestion
// Key concepts, use cases and sign posting
// Explain audit log files, only one active at a time that they’re written etc
```

ElliotFrancisHunter replied:
Just got rid as I'd already fleshed out those points.

simon-dew commented on modules/ROOT/pages/data-modeling.adoc:
```diff
@@ -7,25 +7,22 @@ ifdef::prerelease[:page-status: {prerelease}]
-
 == Introduction
+
 The guidance and constraints documented here relate to the design of data buckets and documents that you require, or may potentially require, to replicate using Sync Gateway functionality.
```
Simplify

```suggestion
This page includes guidance and constraints relating to the design of data buckets and documents that you want to replicate using Sync Gateway.
```

ElliotFrancisHunter replied:
Done. I'm not sure I touched these pages though?

simon-dew commented on modules/ROOT/pages/data-modeling.adoc:
```diff
@@ -62,11 +58,17 @@ This rule applies to writes performed through:
+====
+For versions 2.x of Sync Gateway, you can encounter the following error:
+
+* In Mobile-to-Web Data Sync with Field-level Encryption enabled, because the rule conflicts with the default {xref-sdk-java-encryption}
```
```suggestion
* In Mobile-to-Web Data Sync with Field-level Encryption enabled, because the rule conflicts with the default {xref-sdk-java-encryption}.
```

ElliotFrancisHunter replied:
Done

simon-dew commented on modules/ROOT/pages/manage-audit-logs.adoc:
Still WIP?

ElliotFrancisHunter replied:
Still yes.

simon-dew commented on modules/ROOT/pages/whatsnew.adoc:
```diff
@@ -15,43 +15,15 @@ include::ROOT:partial$_set_page_context.adoc[]
-Couchbase Mobile now offers Scopes and Collections, allowing more efficient and scalable data organisation within a bucket. 
-This also introduces an improved method of defining and enforcing data xref:access-control-concepts.adoc[Access Control] more granularly. 
-Multi-tenant apps will also experience better scalability and independent data lifecycle management.
+Couchbase now provides Audit Logging support. Audit Logging provides tools for administrators to track operational irregularities and to support regulatory and security compliance standards, such as link:https://www.hhs.gov/hipaa/index.html[HIPAA] and link:https://soc2.co.uk/soc2[SOC-2].
```
Specify that it's audit logging support for Sync Gateway?

ElliotFrancisHunter replied:
Done

simon-dew reviewed (dismissed):
Sorry there are such a lot of comments. Many of them are just consistency of punctuation. Also I have some questions about the REST API specs that can't really be fixed in the context of this PR but will have to be looked at later.

ElliotFrancisHunter committed:
DOC-12304: Feedback iteration v1

Updates to docs pages from dev and doc.

Still need to finishing manage audit logs page as there are still notes there. Commit is mainly to provide up to date preview

ElliotFrancisHunter committed:
DOC-12304: Manage logging updates

torcolvin commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -47,100 +47,107 @@ You can alter the configuration per Sync Gateway database by changing the databa
+                        “domain”: “sgw”
+                      }
+                    ],
+               },
```
This json isn't valid, can you pop this into a validator?

1. there is a trailing comma on the last element
2. there are smart quotes on the keys
3. there's missing a `}` to close `"logging"`

ElliotFrancisHunter replied:
Done. Forgot how fiddly JSON validation was!

ElliotFrancisHunter committed:
DOC-12304: Support and Compatibility updates

ElliotFrancisHunter committed:
DOC-12304: Supported versions fixes

Fixed formatting error and removed macOS 11

ElliotFrancisHunter committed:
DOC-12304: Add 3.2 SGW eventing compatibility

ElliotFrancisHunter committed:
DOC-12304: V3 audit logging doc

ElliotFrancisHunter committed:
DOC-12304: Add references to schema

ElliotFrancisHunter committed:
DOC-12304: Fix invalid JSON

ElliotFrancisHunter committed:
DOC-12304: Added 3.1.11 backports to relnotes

osfameron committed:
DOC-12304 update REST API for 3.2

this should also update the database configuration pages.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+[abstract]
+{description}
+
+== Audit Fields 
```
This section is providing the definition for a subset of the properties that may be present on events.  I'm not sure this belongs in the 'Manage Audit Logs' section, as these properties aren't used to configure or manage audit logging.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+
+== Audit Fields 
+
+The table below lists mandatory audit fields with corresponding descriptions. 
```
I'm not sure about the use of the term 'mandatory' here.  Although these are common properties, it's not the case that every audit event emitted includes all of these properties.

ElliotFrancisHunter replied:
Changed!

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+
+IMPORTANT: Audit Logging on Sync Gateway is opt-in functionality and is disabled by default.
+
+Admin users can manage Audit logging via the xref:rest-api-admin.adocp[Admin REST API]. 
```
There are three main ways for users to manage Audit logging:
- bootstrap config (file-based config)
  - global configuration of audit logging
- database config (via existing Admin REST API endpoints)
   - per database configuration of audit logging as part of the database config
- via the new audit-specific Admin REST API endpoint (_config/audit)
    - per database configuration of audit logging only

I think it might make this section easier to follow if this was explained first, before going into the breakdown of each section.  The existing opening sentence ("users can manage Audit logging via the Admin REST API") makes it sound like this is the only way to do it.

ElliotFrancisHunter replied:
Changed

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+IMPORTANT: Audit Logging on Sync Gateway is opt-in functionality and is disabled by default.
+
+Admin users can manage Audit logging via the xref:rest-api-admin.adocp[Admin REST API]. 
+Below is a table of possible configuration options.
```
It's not obvious which configuration option these properties are intended to apply to, or what they correspond to in the config file or REST API payloads.  I'd suggest removing this table and instead providing specific definitions for each of the three configuration options.

The properties in this table also don't align with the implemented product - I think switching to documenting the bootstrap config, database config and audit REST endpoint will avoid that problem

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+"logging": {
+    "audit": {
+	  “audit_log_file_path”: “/path/to/file.log”,
+    “enabled”: true,
```
```suggestion
           “enabled”: true
```

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+
+NOTE: Configuration via the bootstrap config will affect your configuration globally.
+
+To successfully configure audit logging, you must set a file path for your audit logs and set the `enabled` parameter to `true`.
```
If `audit_log_file_path` is not set, SGW will use the more general logging log file path (set in the bootstrap config)

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+To successfully configure audit logging, you must set a file path for your audit logs and set the `enabled` parameter to `true`.
+Enabling and disabling audit logging is handled by the Admin REST API at a cluster and Sync Gateway database level.
+Setting enabled to true without any specified enabled events will only enable events with `DefaultEnabled` set to `true`. 
+To view the list of audit events, see xref:audit-log-events.adoc[]
```
It doesn't mention here (or elsewhere?) that the audit log will be written to a file named sg_audit.log.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+
+=== Configuration with the Database Config
+
+You can alter the configuration per Sync Gateway database by changing the database config.
```
I think it's worth clarifying the hierarchy here.  By default, the bootstrap audit configuration applies to all databases.  Specifying audit logging configuration in the database config allows users to override the global settings for a particular database.  Given that hierarchy, I wonder if it makes sense to go into the additional details about the bootstrap config first (explaining filtering, etc) before circling back to explaining how these details can be configured at the db level.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+You can alter the following to affect events stored in the log file:
+
+* Enabling and disabling users - This affects all events related to a specified user.
+If a given user is disabled, no audit events from the user will be logged.
```
```suggestion
If a given user is disabled, no audit events associated with the user will be logged.
```
I feel like 'associated with' better describes the scenario than 'from' - 'from' sounds like the user is creating an audit event, when really the audit event is associated with some user action.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+* Enabling and disabling events - This affects the specified events globally.
+If a given event is disabled, it will not be logged regardless of the user.
+
+NOTE: If an audit event has `DefaultEnabled` set to `true`, it does not need to be explicitly listed in the `enabled_events: []` array.
```
I don't think this is correct.  If enabled_events is specified in the bootstrap config, that overrides the default set of enabled events.  Only if enabled_events is omitted will the default set of optional events be enabled.
https://confluence.issues.couchbase.com/wiki/spaces/SWD/pages/2387378237/SGW+3.2.0+Audit+Logging#Bootstrap%2Fstartup-config

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+            }
+----
+
+Enabling the events at runtime without modifying the entire database
```
At the end of the day these endpoints are going to modify the database (by updating the audit config for the database) - so this description isn't exactly right.  It would be better described as something like 'Modifying the database audit logging config independently'

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+
+Enabling the events at runtime without modifying the entire database
+
+You can enable events at runtime without modifying the entire database using the following HTTP request methods:
```
Since 'GET' doesn't enable events, I might describe this as 'You can manage the audit logging configuration for a database....'

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+----
+{
+	“enabled”: true,
+  “enabled_events”: [{“53270”: true, “53280”: true]
```
Name and format of the events property isn't quite right here, and should be indented to match 'enabled'.

```suggestion
        “events”: {“53270”: true, “53280”: true}
```

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+  “logging”: {
+               “audit”: {
+                          "enabled”: true,
+                          “enabled_events”: [],
```
```suggestion
                          “events”: {},
```

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,292 @@
+For more information, see xref:access-control-concepts.adoc#lbl-sgw-users[Sync Gateway defined Users and Roles]. 
+* `cbs` - Users that are are RBAC controlled.
+These are created on Couchbase Server.
+For more information, see xref:access-control-concepts.adoc#lbl-rbac-users[RBAC Users]. 
```
It's probably worth explaining how the Couchbase Server users interact with Sync Gateway.  Sync Gateway's admin and metrics REST API are authenticated using Couchbase Server users.

adamcfraser reviewed (dismissed):
Completed a partial review - will continue once I've had a chance to review Iveta's offline comments.

ElliotFrancisHunter committed:
DOC-12304: Correct JIRA URL

osfameron committed:
DOC-12304 update links for REST API

ElliotFrancisHunter committed:
DOC-12304: Eventing compat tweaks

ElliotFrancisHunter committed:
DOC-12304: Adding missing deprecation lines

ElliotFrancisHunter committed:
DOC-12304: Content simplification and reorg

ElliotFrancisHunter committed:
DOC-12304: P0 additions

ElliotFrancisHunter committed:
DOC-12304: Added audit logging command line options

ElliotFrancisHunter committed:
DOC-12304: Reorder as per feedback for logical grouping

ElliotFrancisHunter committed:
DOC-12304: Copy update

ElliotFrancisHunter committed:
DOC-12304: Restructuring

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,311 @@
+[cols="3*"]
+|===
+
+| Name | Values | Additional Description
```
This table is incorrect and doesn't represent the audit logging options present in the bootstrap configuration.  

Use Audit-logging-config (https://github.com/couchbase/sync_gateway/blob/main/docs/api/components/schemas.yaml#L2737) as the source of truth for bootstrap audit logging config format and properties.

ElliotFrancisHunter replied:
Fixed

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,311 @@
+"logging": {
+    "audit": {
+	  "audit_log_file_path": "/path/to/file.log",
+    "enabled": true,
```
Needs alignment tweaked, "enabled" should be left-aligned with "audit_log_file_path".

simon-dew reviewed (approved):
Took a quick look at the new Manage Audit Logs page, LGTM 👍

ElliotFrancisHunter committed:
DOC-12304: Addressing dev feedback

ElliotFrancisHunter committed:
DOC-12304: PM feedback

adamcfraser commented on modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-0-release-note.adoc:
```diff
@@ -0,0 +1,215 @@
+--
+
+Do not deploy Eventing/Sync Gateway until all SGW nodes are at version 3.2.
+For older Sync Gateway versions that do not write import XATTRs, Eventing functions will encounter infinite recursions and duplicate mutations if deployed in a mixed mode SGW environment.
```
```suggestion
For older Sync Gateway versions that do not write import XATTRs, Eventing functions may encounter infinite recursions and duplicate mutations if deployed in a mixed mode SGW environment.
```

ElliotFrancisHunter replied:
Done

adamcfraser commented on modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-0-release-note.adoc:
```diff
@@ -0,0 +1,215 @@
+
+--
+
+Version 3.2.0 of Sync Gateway delivers the following features and enhancements:
```
The main release-defining feature in 3.2.0 (Audit Logging) is sort of buried in a long list of enhancements here.  I think in other releases we've highlighted the larger features at the top of the release notes in a 'New Features' section, possibly with a link to the What's New page?
e.g. https://docs.couchbase.com/sync-gateway/3.0/release-notes.html#lbl-rel300

adamcfraser commented on modules/ROOT/pages/_partials/release_notes/sync-gateway-3-2-0-release-note.adoc:
```diff
@@ -0,0 +1,215 @@
+[#maint-3-2-0]
+=== Fixed Issues
+
+* https://jira.issues.couchbase.com/browse/CBG-2905[CBG-2905 -  Fixed repeated attempts to connect to inexistent buckets are done after removing a database and bucket]
```
Is it standard to add the word 'Fixed' to most of the fixed issues descriptions as you've done here?  I thought that might be inferred from the fact that these are in the 'Fixed Issues' section.

ElliotFrancisHunter replied:
The aim is to make the tickets read more as full sentences on the documentation side

adamcfraser commented on modules/ROOT/pages/_partials/sgw-svr-compatibility.adoc:
```diff
@@ -192,6 +205,17 @@ Compatibility Matrix
 | image:ROOT:no.png[]
 | image:ROOT:yes.png[]
 | image:ROOT:yes.png[]
+| image:ROOT:yes.png[]
```
There's no difference between 3.1 and 3.2 from a compatibility perspective, so I think that means we should be specifying '3.2.0' and '3.2.0 Using Scopes and Collections' rows that match 3.1.0.

ElliotFrancisHunter replied:
Done

adamcfraser commented on modules/ROOT/pages/whatsnew.adoc:
```diff
@@ -15,43 +15,30 @@ include::ROOT:partial$_set_page_context.adoc[]
-* *Collection-Level Sync Functions and Scoped User Associations* +
-The xref:sync-function.adoc[Sync Functions] now work on a Collection level, and additional optional fields have been added to the database configurations to support this update. 
-Each database is designed to support only one Scope. Users can be associated with that Scope and shared across multiple Collections.
+Sync Gateway 3.2.0 now supports interoperability with Eventing from Couchbase Server version 7.6.3+.
```
```suggestion
Sync Gateway 3.2.0 now supports improved interoperability with Eventing from Couchbase Server version 7.6.3+.
```

ElliotFrancisHunter replied:
Done

adamcfraser commented on modules/ROOT/pages/whatsnew.adoc:
```diff
@@ -15,43 +15,30 @@ include::ROOT:partial$_set_page_context.adoc[]
-If you have an app that uses bucket-based APIs, you can still upgrade to 3.1, but please note that this API is now deprecated. 
-For more information, see xref:scopes-and-collections-config.adoc[Scopes and Collections Configuration for Sync Gateway].
+Do not deploy Eventing/Sync Gateway until all SGW nodes are at version 3.2.
+For older Sync Gateway versions that do not write import XATTRs, Eventing functions will encounter infinite recursions and duplicate mutations if deployed in a mixed mode SGW environment.
```
```suggestion
For older Sync Gateway versions that do not write import XATTRs, Eventing functions may encounter infinite recursions and duplicate mutations if deployed in a mixed mode SGW environment.
```

ElliotFrancisHunter replied:
Done!

adamcfraser reviewed (dismissed):
Caught a few more things on this pass - thanks.

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,316 @@
+[source, json, indent=0]
+----
+
+"logging": {
+    "audit": {
+                "audit_log_file_path": "/path/to/file.log",
+                "enabled": true,
+             }
+}
```
```suggestion
"logging": {
    "audit": {
        "audit_log_file_path": "/path/to/file.log",
        "enabled": true
    }
}
```

adamcfraser commented on modules/ROOT/pages/manage-audit-logs.adoc:
```diff
@@ -0,0 +1,316 @@
+.Response
+[source,json]
+----
+{
+
+    "enabled": true,
+    "events": 
+{
+
+    "53270": true,
+    "53271": true
+
+},
+"disabled_users": 
+[
+
+    {
+        "domain": "cbs",
+        "name": "user1"
+    }
+
+],
+"disabled_roles": 
+[
+
+        {
+            "domain": "sgw",
+            "name": "role1"
+        }
+    ]
+
+}
```
```suggestion
{
    "enabled": true,
    "events": {
        "53270": true,
        "53271": true
    },
    "disabled_users": [
        {
            "domain": "cbs",
            "name": "user1"
        }
    ],
    "disabled_roles": [
        {
            "domain": "sgw",
            "name": "role1"
        }
    ]
}
```

adamcfraser reviewed (commented):
Additional json formatting/syntax fixes.

ElliotFrancisHunter committed:
Update modules/ROOT/pages/manage-audit-logs.adoc

Co-authored-by: Adam Fraser <adam.fraser@couchbase.com>

ElliotFrancisHunter committed:
Update modules/ROOT/pages/manage-audit-logs.adoc

Co-authored-by: Adam Fraser <adam.fraser@couchbase.com>

ElliotFrancisHunter committed:
DOC-12304: Final dev feedback before GA

ElliotFrancisHunter committed:
DOC-12304: Content context shuffling

ElliotFrancisHunter committed:
DOC-12304: Specify xref text to remove app services

ElliotFrancisHunter committed:
DOC-12304: Typo fix and condense into one 'fields' column
